src/gui/cookiesmodel.cpp

```diff
@@ -0,0 +1,173 @@
+ * exception statement from your version.
+ */
+
+#include <QDateTime>
+#include "cookiesmodel.h"
+
+CookiesModel::CookiesModel(const QList<QNetworkCookie> &cookies, QObject *parent)
+    : QAbstractItemModel(parent)
+    , m_cookies(cookies)
+{
+}
+
+QVariant CookiesModel::headerData(int section, Qt::Orientation orientation, int role) const
+{
+    if ((role == Qt::DisplayRole) && (orientation == Qt::Horizontal)) {
+        switch (section)
+        {
+        case COL_DOMAIN:
+            return QString("Domain");
+        case COL_PATH:
+            return QString("Path");
+        case COL_NAME:
+            return QString("Name");
+        case COL_VALUE:
+            return QString("Value");
+        case COL_EXPDATE:
+            return QString("Expiration Date");
+        }
+    }
+
+    return QVariant();
+}
+
+QModelIndex CookiesModel::index(int row, int column, const QModelIndex &parent) const
+{
+    if (parent.isValid() // no items with valid parent
+        || (row < 0) || (row >= m_cookies.size())
+        || (column < 0) || (column >= NB_COLUMNS))
+        return QModelIndex();
+
+    return createIndex(row, column, &m_cookies[row]);
+}
+
+QModelIndex CookiesModel::parent(const QModelIndex &index) const
+{
+    Q_UNUSED(index);
+    return QModelIndex();
+}
+
+int CookiesModel::rowCount(const QModelIndex &parent) const
+{
+    if (parent.isValid()) return 0;
+
+    return m_cookies.size();
+}
+
+int CookiesModel::columnCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent);
+    return NB_COLUMNS;
+}
+
+QVariant CookiesModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || (index.row() >= m_cookies.size())
+        || ((role != Qt::DisplayRole) && (role != Qt::EditRole)))
+        return QVariant();
+
+    switch (index.column()) {
+    case COL_DOMAIN:
+        return m_cookies[index.row()].domain();
+    case COL_PATH:
+        return m_cookies[index.row()].path();
+    case COL_NAME:
+        return m_cookies[index.row()].name();
+    case COL_VALUE:
+        return m_cookies[index.row()].value();
+    case COL_EXPDATE:
+        return m_cookies[index.row()].expirationDate();
+    }
+
+    return QVariant();
+}
+
+bool CookiesModel::setData(const QModelIndex &index, const QVariant &value, int role)
+{
+    if (role != Qt::EditRole) return false;
+
+    switch (index.column()) {
+    case COL_DOMAIN:
+        m_cookies[index.row()].setDomain(value.toString());
+        break;
+    case COL_PATH:
+        m_cookies[index.row()].setPath(value.toString());
+        break;
+    case COL_NAME:
+        m_cookies[index.row()].setName(value.toString().toUtf8());
+        break;
+    case COL_VALUE:
+        m_cookies[index.row()].setValue(value.toString().toUtf8());
+        break;
+    case COL_EXPDATE:
+        m_cookies[index.row()].setExpirationDate(value.toDateTime());
+        break;
+    default:
+        return false;
+    }
```

Comment:
Hmm, I am confused. Why do you need to convert to utf8? You copy one QString to another QString. They use the same internal representation already.